src/fs/fuse.rs

```diff
@@ -38,1190 +38,557 @@ const U64_SIZE: usize = ::core::mem::size_of::<u64>();
+/// Marker trait, which signals that a struct is a valid Fuse response.
+/// Struct has to be repr(C)!
+pub(crate) unsafe trait FuseOut {}
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_init_in {
-	pub major: u32,
-	pub minor: u32,
-	pub max_readahead: u32,
-	pub flags: u32,
+unsafe impl FuseIn for fuse_abi::InitIn {}
+unsafe impl FuseOut for fuse_abi::InitOut {}
+unsafe impl FuseIn for fuse_abi::ReadIn {}
+unsafe impl FuseIn for fuse_abi::WriteIn {}
+unsafe impl FuseOut for fuse_abi::WriteOut {}
+unsafe impl FuseOut for fuse_abi::ReadOut {}
+unsafe impl FuseIn for fuse_abi::LookupIn {}
+unsafe impl FuseIn for fuse_abi::ReadlinkIn {}
+unsafe impl FuseOut for fuse_abi::ReadlinkOut {}
+unsafe impl FuseOut for fuse_abi::AttrOut {}
+unsafe impl FuseOut for fuse_abi::EntryOut {}
+unsafe impl FuseIn for fuse_abi::CreateIn {}
+unsafe impl FuseOut for fuse_abi::CreateOut {}
+unsafe impl FuseIn for fuse_abi::OpenIn {}
+unsafe impl FuseOut for fuse_abi::OpenOut {}
+unsafe impl FuseIn for fuse_abi::ReleaseIn {}
+unsafe impl FuseOut for fuse_abi::ReleaseOut {}
+unsafe impl FuseIn for fuse_abi::RmdirIn {}
+unsafe impl FuseOut for fuse_abi::RmdirOut {}
+unsafe impl FuseIn for fuse_abi::MkdirIn {}
+unsafe impl FuseIn for fuse_abi::UnlinkIn {}
+unsafe impl FuseOut for fuse_abi::UnlinkOut {}
+unsafe impl FuseIn for fuse_abi::LseekIn {}
+unsafe impl FuseOut for fuse_abi::LseekOut {}
+unsafe impl FuseIn for fuse_abi::PollIn {}
+unsafe impl FuseOut for fuse_abi::PollOut {}
+
+pub(crate) trait OpTrait {
+	type InStruct: FuseIn + core::fmt::Debug;
+	type InPayload: ?Sized;
+	type OutStruct: FuseOut + core::fmt::Debug;
+	type OutPayload: ?Sized;
+
+	type Cmd: ?Sized + AsSliceU8 = ReqPart<fuse_abi::InHeader, Self::InStruct, Self::InPayload>;
+	type Rsp: ?Sized + AsSliceU8 =
+		ReqPart<fuse_abi::OutHeader, MaybeUninit<Self::OutStruct>, Self::OutPayload>;
 }
 
-unsafe impl FuseIn for fuse_init_in {}
+pub(crate) struct Op<const CODE: u32>;
 
-#[repr(C)]
-#[derive(Debug, Default)]
-struct fuse_init_out {
-	pub major: u32,
-	pub minor: u32,
-	pub max_readahead: u32,
-	pub flags: u32,
-	pub max_background: u16,
-	pub congestion_threshold: u16,
-	pub max_write: u32,
-	pub time_gran: u32,
-	pub unused: [u32; 9],
+impl OpTrait for Op<{ fuse_abi::Opcode::Init as u32 }> {
+	type InStruct = fuse_abi::InitIn;
+	type InPayload = ();
+	type OutStruct = fuse_abi::InitOut;
+	type OutPayload = ();
 }
-unsafe impl FuseOut for fuse_init_out {}
 
-#[repr(C)]
-#[derive(Default, Debug)]
-struct fuse_read_in {
-	pub fh: u64,
-	pub offset: u64,
-	pub size: u32,
-	pub read_flags: u32,
-	pub lock_owner: u64,
-	pub flags: u32,
-	pub padding: u32,
+impl Op<{ fuse_abi::Opcode::Init as u32 }> {
+	fn create() -> (Box<<Self as OpTrait>::Cmd>, Box<<Self as OpTrait>::Rsp>) {
+		let cmd = Self::new_cmd(
+			fuse_abi::ROOT_ID,
+			fuse_abi::InitIn {
+				major: 7,
+				minor: 31,
+				max_readahead: 0,
+				flags: 0,
+			},
+		);
+		let rsp = unsafe { Box::new_uninit().assume_init() };
```

Comment:
This is immediate undefined behavior. You need to either initialize `rsp` here, or return `MaybeUninit`.

This applies to all other instances of `new_uninit().assume_init()` too (except for producing, say `[MaybeUninit<u8>; N]` or other uninitialized structs). See [`std::mem::uninitialized`](https://doc.rust-lang.org/nightly/std/mem/fn.uninitialized.html) for details on why this is bad.

Reply:
I tried using MaybeUninit but it did not work because `Rsp` is a dynamically-sized type (at least in the cases where the output payload is not `()`). What would be an appropriate alternative here?

Reply:
For `OutPayload = ()`, we could `impl zerocopy::FromZeroes for Op` and use `FromZeros::new_box_zeroed()`. For `OutPayload = [MaybeUninit<u8>]` we have to at least initialize the non-`MaybeUninit` part.

BTW: In the future, [`DynSized`](https://github.com/rust-lang/rfcs/pull/3536) hopefully improves on all of this. :D

Reply:
Would it be safe to `assume_init` the struct if the fields inside the struct are MaybeUninit? I would assume it would be since it is not possible to access the fields, but I don't know for sure.

Otherwise, I think init the struct with fields set to `MaybeUninit::uninit()`s, but I think the first option is preferable since it saves us from writing a bunch of struct literals with `MaybeUninit::uninit()`s.

Reply:
> Would it be safe to `assume_init` the struct if the fields inside the struct are MaybeUninit?

Yeah, that would correspond to the `[MaybeUninit<u8>; N]` case above.

Reply:
The new definition for Rsp declares each field as MaybeUninit (or (), in the case of payloadless Rsps).